src/chainstate/burn/db/sortdb.rs

```diff
@@ -591,19 +609,221 @@ impl <'a> SortitionHandleTx <'a> {
             return Err(db_error::ReadOnly);
         }
 
-        let tx = tx_begin_immediate(&mut conn.conn)?;
         let handle = SortitionHandleTx::new(
-            tx, &mut conn.marf,
+            &mut conn.marf,
             SortitionHandleContext { chain_tip: parent_chain_tip.clone(),
                                      first_block_height: conn.first_block_height });
 
         Ok(handle)
     }
 
-    pub fn conn_view<'b>(&'b self, view_sortition: &SortitionId) -> SortitionHandleConn<'b> {
-        let mut conn_view = self.as_conn();
-        conn_view.context.chain_tip = view_sortition.clone();
-        conn_view
+    /// Uses the handle's current fork identifier to get a block snapshot by
+    ///   burnchain block header
+    /// If the burn header hash is _not_ in the current fork, then this will return Ok(None)
+    pub fn get_block_snapshot(&mut self, burn_header_hash: &BurnchainHeaderHash, chain_tip: &SortitionId) -> Result<Option<BlockSnapshot>, db_error> {
+        let sortition_identifier_key = db_keys::sortition_id_for_bhh(burn_header_hash);
+        let sortition_id = match self.get_indexed(&chain_tip, &sortition_identifier_key)? {
+            None => return Ok(None),
+            Some(x) => SortitionId::from_hex(&x).expect("FATAL: bad Sortition ID stored in DB")
+        };
+
+        SortitionDB::get_block_snapshot(self.tx(), &sortition_id)
+    }
+
+    /// Get a leader key at a specific location in the burn chain's fork history, given the
+    /// matching block commit's fork index root (block_height and vtxindex are the leader's
+    /// calculated location in this fork).
+    /// Returns None if there is no leader key at this location.
+    pub fn get_leader_key_at(&mut self, key_block_height: u64, key_vtxindex: u32, tip: &SortitionId) -> Result<Option<LeaderKeyRegisterOp>, db_error> {
+        assert!(key_block_height < BLOCK_HEIGHT_MAX);
+        let ancestor_snapshot = match SortitionDB::get_ancestor_snapshot_tx(self, key_block_height, tip)? {
+            Some(sn) => sn,
+            None => {
+                return Ok(None);
+            }
+        };
+
+        // PoX TODO: note -- leader_keys table will index on burn_header_hash: if a leader key is reprocessed due to a PoX fork,
```

Comment:
It shouldn't matter because (burn_header_hash, block_height, vtxindex) will always identify the same transaction, but I updated it to use `sortition_id`. This method shouldn't check `pox_valid`, because it is already getting called from the context of a specific "sortition fork" -- whether or not that fork is a valid fork isn't really a concern of this function.